core/trino-main/src/main/java/io/trino/operator/PartitionedLookupSource.java

```diff
@@ -150,6 +152,70 @@ public long getJoinPosition(int position, Page hashChannelsPage, Page allChannel
         return encodePartitionedJoinPosition(partition, toIntExact(joinPosition));
     }
 
+    @Override
+    public long[] getJoinPositions(int[] positions, Page hashChannelsPage, Page allChannelsPage)
+    {
+        if (!supportsCaching()) {
+            return LookupSource.super.getJoinPositions(positions, hashChannelsPage, allChannelsPage);
+        }
+
+        long[] rawHashes = new long[positions.length];
+        for (int i = 0; i < positions.length; i++) {
+            rawHashes[i] = partitionGenerator.getRawHash(hashChannelsPage, positions[i]);
+        }
+
+        return getJoinPositions(positions, hashChannelsPage, allChannelsPage, rawHashes);
+    }
+
+    @Override
+    public long[] getJoinPositions(int[] positions, Page hashChannelsPage, Page allChannelsPage, long[] rawHashes)
+    {
+        if (!supportsCaching()) {
+            return LookupSource.super.getJoinPositions(positions, hashChannelsPage, allChannelsPage, rawHashes);
+        }
+        long[] result = new long[positions.length];
+
+        // Determine the partition of each element of positions array
+        int[] partitions = new int[positions.length];
```

Comment:
Why not to use `IntArrayList[]` as in `PartitioningExchanger`. That would save one pass and extra allocations